### ISSUE TITLE
Dynamic endpoint cannot be resolved properly (toD)

### ISSUE THREAD

**Issue by hakuseki** [apache/camel-quarkus#1116]
Using latest version of Camel Quarkus I cannot get a `toD()` with dynamic http endpoints to be resolved correctly.

` .toD("{{esma.download.url}}")`

which is resolved to this:

`https://registers.esma.europa.eu/solr/esma_registers_firds_files/select?q=*&fq=publication_date:%5B${date:now-72h:yyyy-MM-dd}T00:00:00Z%20TO%20${date:now-24h:yyyy-MM-dd}T23:59:59Z%5D&wt=xml&indent=false&start=0&rows=100`

But the dynamic values (a date calculation) is not resolved.

I have tried to debug it and I don't really grasp how and where but the findings so far is that in class `SendDynamicProcessor.java` there's a property called `dynamicAware` that doesn't get resolved (null). The reason here is that is searches for a path
`META-INF/services/org/apache/camel/send-dynamic/https` in a `classMap` which is not found (`FastFactoryResolver.java`).

**Comment by lburgazzoli** [comment]
can you try to add

    quarkus.camel.service.discovery.include-patterns = META-INF/services/org/apache/camel/send-dynamic/http

to the application.properties ?

**Comment by davsclaus** [comment]
Oh there is a few components that has this - especially the HTTP based so we should find a solution for this in camel-quarkus

**Comment by lburgazzoli** [comment]
Yeah we can add this to the default list of locations, I just need a confirmation that this fixes the issue, then I can fix it

**Comment by hakuseki** [comment]
inner issue fixed but then I get a NPE in

`2020-04-21 15:52:29,888 WARN  [org.apa.cam.com.tim.TimerConsumer] (Camel (camel-2) thread #10 - timer://FIRDSDownloader) Error processing exchange. Exchange[93CF35C1B474473-0000000000000000]. Caused by: [java.lang.NullPointerException - null]: java.lang.NullPointerException
        at org.apache.camel.catalog.impl.AbstractCamelCatalog.componentJSonSchema(AbstractCamelCatalog.java:71)
        at org.apache.camel.catalog.impl.DefaultRuntimeCamelCatalog.lambda$componentJSonSchema$2(DefaultRuntimeCamelCatalog.java:75)
        at org.apache.camel.catalog.impl.DefaultRuntimeCamelCatalog.cache(DefaultRuntimeCamelCatalog.java:126)
        at org.apache.camel.catalog.impl.DefaultRuntimeCamelCatalog.componentJSonSchema(DefaultRuntimeCamelCatalog.java:75)
        at org.apache.camel.catalog.impl.AbstractCamelCatalog.componentModel(AbstractCamelCatalog.java:84)
        at org.apache.camel.catalog.impl.DefaultRuntimeCamelCatalog.lambda$componentModel$3(DefaultRuntimeCamelCatalog.java:80)
        at org.apache.camel.catalog.impl.DefaultRuntimeCamelCatalog.cache(DefaultRuntimeCamelCatalog.java:126)
        at org.apache.camel.catalog.impl.DefaultRuntimeCamelCatalog.componentModel(DefaultRuntimeCamelCatalog.java:80)
        at org.apache.camel.catalog.impl.AbstractCamelCatalog.endpointProperties(AbstractCamelCatalog.java:382)
        at org.apache.camel.http.base.HttpSendDynamicAware.prepare(HttpSendDynamicAware.java:58)
        at org.apache.camel.processor.SendDynamicProcessor.process(SendDynamicProcessor.java:124)
        at org.apache.camel.processor.errorhandler.RedeliveryErrorHandler$SimpleTask.run(RedeliveryErrorHandler.java:396)
        at org.apache.camel.impl.engine.DefaultReactiveExecutor$Worker.schedule(DefaultReactiveExecutor.java:153)
        at org.apache.camel.impl.engine.DefaultReactiveExecutor.scheduleMain(DefaultReactiveExecutor.java:60)
        at org.apache.camel.processor.Pipeline.process(Pipeline.java:147)
        at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:286)
        at org.apache.camel.component.timer.TimerConsumer.sendTimerExchange(TimerConsumer.java:198)
        at org.apache.camel.component.timer.TimerConsumer$1.run(TimerConsumer.java:78)
        at java.base/java.util.TimerThread.mainLoop(Timer.java:556)
        at java.base/java.util.TimerThread.run(Timer.java:506)

`

Keep on debugging…

**Comment by davsclaus** [comment]
Ah okay that smells like it lacks some camel-core-catalog files that it uses for optimizing the dynamic endpoints when using toD. There is a JIRA ticket upstream at Apache Camel to implement something that is potentially smarter and source code generated specific tailored java code or something.

**Comment by davsclaus** [comment]
You can turn off optimizing toD with this strategy on toD (there is an option to do that).

**Comment by hakuseki** [comment]
Ahh, I see. The NPE is in `DefaultRuntimeCamelCatalog.java`

`public ComponentModel componentModel(String name) {
        return (ComponentModel)this.cache("component-model-" + name, name, (x$0) -> {
            return super.componentModel(x$0);
        });
    }`

**Comment by davsclaus** [comment]
The JIRA ticket is: https://issues.apache.org/jira/browse/CAMEL-14499

**Comment by lburgazzoli** [comment]
@hakuseki I don't have the NPE, mind sharing a little reproducer ?

**Comment by hakuseki** [comment]
`    from(timer("FIRDSDownloader").repeatCount(1))
                .setHeader("CamelHttpMethod", constant("GET"))
//                .toD("registers.esma.europa.eu/solr/esma_registers_firds_files/select?q=*&fq=publication_date:%5B${date:now-72h:yyyy-MM-dd}T00:00:00Z%20TO%20${date:now-24h:yyyy-MM-dd}T23:59:59Z%5D&wt=xml&indent=false&start=0&rows=100&allowOptimisedComponents=false")
                .toD(https("registers.esma.europa.eu/solr/esma_registers_firds_files/select?q=*&fq=publication_date:%5B${date:now-72h:yyyy-MM-dd}T00:00:00Z%20TO%20${date:now-24h:yyyy-MM-dd}T23:59:59Z%5D&wt=xml&indent=false&start=0&rows=100&allowOptimisedComponents=false"))
                .log("${body}")
                .end();
`

This produces this NPE:

2020-04-21 17:45:45,592 WARN  [org.apa.cam.com.tim.TimerConsumer] (Camel (camel-13) thread #65 - timer://FIRDSDownloader) Error processing exchange. Exchange[B32CD34ABD896B2-0000000000000000]. Caused by: [java.lang.NullPointerException - null]: java.lang.NullPointerException
        at org.apache.camel.catalog.impl.AbstractCamelCatalog.componentJSonSchema(AbstractCamelCatalog.java:71)
        at org.apache.camel.catalog.impl.DefaultRuntimeCamelCatalog.lambda$componentJSonSchema$2(DefaultRuntimeCamelCatalog.java:75)
        at org.apache.camel.catalog.impl.DefaultRuntimeCamelCatalog.cache(DefaultRuntimeCamelCatalog.java:126)
        at org.apache.camel.catalog.impl.DefaultRuntimeCamelCatalog.componentJSonSchema(DefaultRuntimeCamelCatalog.java:75)
        at org.apache.camel.catalog.impl.AbstractCamelCatalog.componentModel(AbstractCamelCatalog.java:84)
        at org.apache.camel.catalog.impl.DefaultRuntimeCamelCatalog.lambda$componentModel$3(DefaultRuntimeCamelCatalog.java:80)
        at org.apache.camel.catalog.impl.DefaultRuntimeCamelCatalog.cache(DefaultRuntimeCamelCatalog.java:126)
        at org.apache.camel.catalog.impl.DefaultRuntimeCamelCatalog.componentModel(DefaultRuntimeCamelCatalog.java:80)
        at org.apache.camel.catalog.impl.AbstractCamelCatalog.endpointProperties(AbstractCamelCatalog.java:382)
        at org.apache.camel.http.base.HttpSendDynamicAware.prepare(HttpSendDynamicAware.java:58)
        at org.apache.camel.processor.SendDynamicProcessor.process(SendDynamicProcessor.java:124)
        at org.apache.camel.processor.errorhandler.RedeliveryErrorHandler$SimpleTask.run(RedeliveryErrorHandler.java:396)
        at org.apache.camel.impl.engine.DefaultReactiveExecutor$Worker.schedule(DefaultReactiveExecutor.java:153)
        at org.apache.camel.impl.engine.DefaultReactiveExecutor.scheduleMain(DefaultReactiveExecutor.java:60)
        at org.apache.camel.processor.Pipeline.process(Pipeline.java:147)
        at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:286)
        at org.apache.camel.component.timer.TimerConsumer.sendTimerExchange(TimerConsumer.java:198)
        at org.apache.camel.component.timer.TimerConsumer$1.run(TimerConsumer.java:78)
        at java.base/java.util.TimerThread.mainLoop(Timer.java:556)
        at java.base/java.util.TimerThread.run(Timer.java:506)

**Comment by hakuseki** [comment]
`    from(timer("FIRDSDownloader").repeatCount(1))
                .setHeader("CamelHttpMethod", constant("GET"))
//                .toD("registers.esma.europa.eu/solr/esma_registers_firds_files/select?q=*&fq=publication_date:%5B${date:now-72h:yyyy-MM-dd}T00:00:00Z%20TO%20${date:now-24h:yyyy-MM-dd}T23:59:59Z%5D&wt=xml&indent=false&start=0&rows=100&allowOptimisedComponents=false")
                .toD(https("registers.esma.europa.eu/solr/esma_registers_firds_files/select?q=*&fq=publication_date:%5B${date:now-72h:yyyy-MM-dd}T00:00:00Z%20TO%20${date:now-24h:yyyy-MM-dd}T23:59:59Z%5D&wt=xml&indent=false&start=0&rows=100&allowOptimisedComponents=false"))
                .log("${body}")
                .end();
`

This produces this NPE:

2020-04-21 17:45:45,592 WARN  [org.apa.cam.com.tim.TimerConsumer] (Camel (camel-13) thread #65 - timer://FIRDSDownloader) Error processing exchange. Exchange[B32CD34ABD896B2-0000000000000000]. Caused by: [java.lang.NullPointerException - null]: java.lang.NullPointerException
        at org.apache.camel.catalog.impl.AbstractCamelCatalog.componentJSonSchema(AbstractCamelCatalog.java:71)
        at org.apache.camel.catalog.impl.DefaultRuntimeCamelCatalog.lambda$componentJSonSchema$2(DefaultRuntimeCamelCatalog.java:75)
        at org.apache.camel.catalog.impl.DefaultRuntimeCamelCatalog.cache(DefaultRuntimeCamelCatalog.java:126)
        at org.apache.camel.catalog.impl.DefaultRuntimeCamelCatalog.componentJSonSchema(DefaultRuntimeCamelCatalog.java:75)
        at org.apache.camel.catalog.impl.AbstractCamelCatalog.componentModel(AbstractCamelCatalog.java:84)
        at org.apache.camel.catalog.impl.DefaultRuntimeCamelCatalog.lambda$componentModel$3(DefaultRuntimeCamelCatalog.java:80)
        at org.apache.camel.catalog.impl.DefaultRuntimeCamelCatalog.cache(DefaultRuntimeCamelCatalog.java:126)
        at org.apache.camel.catalog.impl.DefaultRuntimeCamelCatalog.componentModel(DefaultRuntimeCamelCatalog.java:80)
        at org.apache.camel.catalog.impl.AbstractCamelCatalog.endpointProperties(AbstractCamelCatalog.java:382)
        at org.apache.camel.http.base.HttpSendDynamicAware.prepare(HttpSendDynamicAware.java:58)
        at org.apache.camel.processor.SendDynamicProcessor.process(SendDynamicProcessor.java:124)
        at org.apache.camel.processor.errorhandler.RedeliveryErrorHandler$SimpleTask.run(RedeliveryErrorHandler.java:396)
        at org.apache.camel.impl.engine.DefaultReactiveExecutor$Worker.schedule(DefaultReactiveExecutor.java:153)
        at org.apache.camel.impl.engine.DefaultReactiveExecutor.scheduleMain(DefaultReactiveExecutor.java:60)
        at org.apache.camel.processor.Pipeline.process(Pipeline.java:147)
        at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:286)
        at org.apache.camel.component.timer.TimerConsumer.sendTimerExchange(TimerConsumer.java:198)
        at org.apache.camel.component.timer.TimerConsumer$1.run(TimerConsumer.java:78)
        at java.base/java.util.TimerThread.mainLoop(Timer.java:556)
        at java.base/java.util.TimerThread.run(Timer.java:506)

**Comment by lburgazzoli** [comment]
@hakuseki I've created [this PR](https://github.com/apache/camel-quarkus/pull/1117) with the fix for the first issue but I'm unable to reproduce the NPE

**Comment by hakuseki** [comment]
Thx

So what you are saying is that you can run my route and get a reply from the URL without any NPE?

The URL endpoint is working, tried to use a static URL on a `.to()` and I got a response back.

**Comment by lburgazzoli** [comment]
This is what I do: https://github.com/apache/camel-quarkus/pull/1117/files#diff-26b140eadb8c00c2a78ff3d7e4001450

**Comment by hakuseki** [comment]
Looks like "mine" but I'm curious if I have some dependencies missing or any other structural errors in my code…

**Comment by davsclaus** [comment]
btw allowOptimisedComponents=false is wrong, as its an option on toD itself and not http component. https://camel.apache.org/components/latest/eips/toD-eip.html

However setting this option in Java DSL is not so easy.

the toD docs is a bit wrong with the cacheSize example later on - I will fix that

**Comment by davsclaus** [comment]
And a ticket to make configuring toD easier in Java DSL
https://issues.apache.org/jira/browse/CAMEL-14943

**Comment by hakuseki** [comment]
I have now created a empty project from code.quarkus.io and then added my dependencies and the Download route.

Running it gives me the same NPE:

```
2020-04-22 12:17:44,446 WARN  [org.apa.cam.com.tim.TimerConsumer] (Camel (camel-5) thread #1 - timer://FIRDSDownloader) Error processing exchange. Exchange[52702C36318CF94-0000000000000000]. Caused by: [java.lang.NullPointerException - null]: java.lang.NullPointerException
        at org.apache.camel.catalog.impl.AbstractCamelCatalog.componentJSonSchema(AbstractCamelCatalog.java:71)
        at org.apache.camel.catalog.impl.DefaultRuntimeCamelCatalog.lambda$componentJSonSchema$2(DefaultRuntimeCamelCatalog.java:75)
        at org.apache.camel.catalog.impl.DefaultRuntimeCamelCatalog.cache(DefaultRuntimeCamelCatalog.java:126)
        at org.apache.camel.catalog.impl.DefaultRuntimeCamelCatalog.componentJSonSchema(DefaultRuntimeCamelCatalog.java:75)
        at org.apache.camel.catalog.impl.AbstractCamelCatalog.componentModel(AbstractCamelCatalog.java:84)
        at org.apache.camel.catalog.impl.DefaultRuntimeCamelCatalog.lambda$componentModel$3(DefaultRuntimeCamelCatalog.java:80)
        at org.apache.camel.catalog.impl.DefaultRuntimeCamelCatalog.cache(DefaultRuntimeCamelCatalog.java:126)
        at org.apache.camel.catalog.impl.DefaultRuntimeCamelCatalog.componentModel(DefaultRuntimeCamelCatalog.java:80)
        at org.apache.camel.catalog.impl.AbstractCamelCatalog.endpointProperties(AbstractCamelCatalog.java:382)
        at org.apache.camel.http.base.HttpSendDynamicAware.prepare(HttpSendDynamicAware.java:58)
        at org.apache.camel.processor.SendDynamicProcessor.process(SendDynamicProcessor.java:124)
        at org.apache.camel.processor.errorhandler.RedeliveryErrorHandler$SimpleTask.run(RedeliveryErrorHandler.java:396)
        at org.apache.camel.impl.engine.DefaultReactiveExecutor$Worker.schedule(DefaultReactiveExecutor.java:153)
        at org.apache.camel.impl.engine.DefaultReactiveExecutor.scheduleMain(DefaultReactiveExecutor.java:60)
        at org.apache.camel.processor.Pipeline.process(Pipeline.java:147)
        at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:286)
        at org.apache.camel.component.timer.TimerConsumer.sendTimerExchange(TimerConsumer.java:198)
        at org.apache.camel.component.timer.TimerConsumer$1.run(TimerConsumer.java:78)
        at java.base/java.util.TimerThread.mainLoop(Timer.java:556)
        at java.base/java.util.TimerThread.run(Timer.java:506)
```

**Comment by lburgazzoli** [comment]
can you upload the project to github ?

**Comment by hakuseki** [comment]
https://github.com/hakuseki/CamelQuarkusDynamicHttp

**Comment by hakuseki** [comment]
removed all jdbc/flyway stuff for convenience

**Comment by hakuseki** [comment]
Expected result would be:

```
<?xml version="1.0" encoding="UTF-8"?>
<response>

<lst name="responseHeader">
  <int name="status">0</int>
  <int name="QTime">0</int>
  <lst name="params">
    <str name="q">*</str>
    <str name="indent">false</str>
    <str name="start">0</str>
    <str name="fq">publication_date:[2020-04-21T00:00:00Z TO 2020-04-21T23:59:59Z]</str>
    <str name="rows">100</str>
    <str name="wt">xml</str>
  </lst>
</lst>
<result name="response" numFound="2" start="0">
  <doc>
    <date name="publication_date">2020-04-21T00:00:00Z</date>
    <str name="download_link">http://firds.esma.europa.eu/firds/DLTINS_20200421_01of02.zip</str>
    <str name="file_type">DLTINS</str>
    <str name="file_name">DLTINS_20200421_01of02.zip</str>
    <str name="id">34659</str>
    <int name="_root_">34659</int>
    <str name="published_instrument_file_id">34659</str>
    <long name="_version_">1664650793854173187</long>
    <date name="timestamp">2020-04-22T05:51:47.884Z</date></doc>
  <doc>
    <date name="publication_date">2020-04-21T00:00:00Z</date>
    <str name="download_link">http://firds.esma.europa.eu/firds/DLTINS_20200421_02of02.zip</str>
    <str name="file_type">DLTINS</str>
    <str name="file_name">DLTINS_20200421_02of02.zip</str>
    <str name="id">34660</str>
    <int name="_root_">34660</int>
    <str name="published_instrument_file_id">34660</str>
    <long name="_version_">1664650793854173188</long>
    <date name="timestamp">2020-04-22T05:51:47.884Z</date></doc>
</result>
</response>

```

**Comment by lburgazzoli** [comment]
ok so I've been able to narrow down the issue and it has been solved in camel 3.2.0 so it will be available from the next camel-quarkus release
